Fix: Pass data, fromPlugins, toPlugins, originalPlugins through journal rather than separately (part of #4)

joe-allen-89 opened the pull request:
Part of #4 

Todo: Create journal entry for plugin changes

oliverfoster commented:
Almost there. They should be be passed into the journal as:
```js
            const journal = new Journal({
              data: { 
                data,
                fromPlugins
              },
              supplementEntry: (entry, data) => {
                entry._id = data[entry.keys[0]]?._id ?? '';
                entry._type = data[entry.keys[0]]?._type ?? '';
                if (entry._type && data[entry.keys[0]]?.[`_${entry._type}`]) {
                  entry[`_${entry._type}`] = data[entry.keys[0]]?.[`_${entry._type}`] ?? '';
                }
                return entry;
              }
            });
```

And extracted with:

```js
    this.context = new TaskContext({
      fromPlugins,
      originalFromPlugins,
      toPlugins,
      fromPlugins: journal.subject.fromPlugins,
      originalFromPlugins: journal.subject.originalFromPlugins,
      toPlugins: journal.subject.toPlugins,
      data: journal.subject.data,
      journal
    })
```

Only changes to the `data` property of the `Journal` instance are kept as journal entries.

joe-allen-89 commented:
> Almost there. They should be be passed into the journal as:
> 
> ```js
>             const journal = new Journal({
>               data: { 
>                 data,
>                 fromPlugins
>               },
>               supplementEntry: (entry, data) => {
>                 entry._id = data[entry.keys[0]]?._id ?? '';
>                 entry._type = data[entry.keys[0]]?._type ?? '';
>                 if (entry._type && data[entry.keys[0]]?.[`_${entry._type}`]) {
>                   entry[`_${entry._type}`] = data[entry.keys[0]]?.[`_${entry._type}`] ?? '';
>                 }
>                 return entry;
>               }
>             });
> ```
> 
> And extracted with:
> 
> ```js
>     this.context = new TaskContext({
>       fromPlugins: journal.subject.fromPlugins,
>       originalFromPlugins: journal.subject.originalFromPlugins,
>       toPlugins: journal.subject.toPlugins,
>       data: journal.subject.data,
>       journal
>     })
> ```
> 
> Only changes to the `data` property of the `Journal` instance are kept as journal entries. It's this line where the magic happens partly
> 
> https://github.com/cgkineo/adapt-migrations/blob/b2419e6e30597380387798b3b16723a4c2d1b055/lib/Journal.js#L209



> Almost there. They should be be passed into the journal as:
> 
> ```js
>             const journal = new Journal({
>               data: { 
>                 data,
>                 fromPlugins
>               },
>               supplementEntry: (entry, data) => {
>                 entry._id = data[entry.keys[0]]?._id ?? '';
>                 entry._type = data[entry.keys[0]]?._type ?? '';
>                 if (entry._type && data[entry.keys[0]]?.[`_${entry._type}`]) {
>                   entry[`_${entry._type}`] = data[entry.keys[0]]?.[`_${entry._type}`] ?? '';
>                 }
>                 return entry;
>               }
>             });
> ```
> 
> And extracted with:
> 
> ```js
>     this.context = new TaskContext({
>       fromPlugins: journal.subject.fromPlugins,
>       originalFromPlugins: journal.subject.originalFromPlugins,
>       toPlugins: journal.subject.toPlugins,
>       data: journal.subject.data,
>       journal
>     })
> ```
> 
> Only changes to the `data` property of the `Journal` instance are kept as journal entries. It's this line where the magic happens partly
> 
> https://github.com/cgkineo/adapt-migrations/blob/b2419e6e30597380387798b3b16723a4c2d1b055/lib/Journal.js#L209

Passed in as an object, to have everything as part of the subject I've updated the journal to wrap this.data and condition to the return early on addEntry so only data changes will be added.

oliverfoster commented on lib/Journal.js:
```diff
@@ -183,12 +183,16 @@ export default class Journal {
      * @type {Object|Array}
      * */
-    this.data = data
+    this.data = {
+      ...data,
+      originalFromPlugins: JSON.parse(JSON.stringify(data.fromPlugins)),
+      toPlugins: JSON.parse(JSON.stringify(data.fromPlugins))
+    }
```
this stuff should be done outside of the journal as the journal is only for doing journal things on the data

oliverfoster commented on lib/Journal.js:
```diff
@@ -79,7 +79,7 @@ export default class Journal {
     const addEntry = (method, keys, value = null, previous = null) => {
       // Prevent entries from being added whilst undoing.
-      if (this.isUndoing) return
+      if (this.isUndoing || keys[0] !== 'data') return
```
Yes, I want all of the plugin changes and data changes represented in the journal entries, such that if the journal entries are read, it's clear how the plugins have been updated and how the data has changed accordingly.

joe-allen-89 replied:
Misunderstanding on my part when you said only changes to data. The other data. As it's no longer the only data maybe we change 'data' to 'content'?

oliverfoster replied:
Yup, perfect.

oliverfoster commented on README.md:
```diff
@@ -179,9 +179,15 @@ module.exports = function(grunt) {
+              data: {
+                data,
+                fromPlugins,
+                originalFromPlugins: originalFromPlugins,
```
```suggestion
                originalFromPlugins,
```

joe-allen-89 replied:
I've changed it to originalFromPlugins: originalFromPlugins as I've moved 'originalFromPlugins = JSON.parse(JSON.stringify(fromPlugins))' from taskContext to the grunt task as it seemed to make more sense setting it before the journal is set up rather than part way through, happy to change back though?

oliverfoster replied:
They are functionally identical, `originalFromPlugins,` is just a shorthand for `originalFromPlugins: originalFromPlugins,`.

oliverfoster commented on lib/Task.js:
```diff
@@ -284,13 +284,14 @@ export default class Task {
         const journal = new Journal({
-          data
+          data,
+          fromPlugins
```
Same needs to happen in test execution.

oliverfoster commented:
Once those last bits are in, we probably need to ask if the task stuff needs to know about the journal at all, or if we can just pass through the required journal parts, `journal.subject.*`. 

Really the journal is just providing a version of `fromPlugins`, `content` etc, where the changes can be recorded. 

I think I'm happy with passing the journal around as a single variable, just about. It does make them intent of each receiving function a little less clear though, where having a variable each is explicit about intent.

It would be more clear or more ideal if they were separated and the task stuff new nothing off the journal bits.

I'd need to identify if there are any parts of the task classes that needs to use any of the journal APIs first.

What do you think?

joe-allen-89 commented:
> Once those last bits are in, we probably need to ask if the task stuff needs to know about the journal at all, or if we can just pass through the required journal parts, `journal.subject.*`.
> 
> Really the journal is just providing a version of `fromPlugins`, `content` etc, where the changes can be recorded.
> 
> I think I'm happy with passing the journal around as a single variable, just about. It does make them intent of each receiving function a little less clear though, where having a variable each is explicit about intent.
> 
> It would be more clear or more ideal if they were separated and the task stuff new nothing off the journal bits.
> 
> I'd need to identify if there are any parts of the task classes that needs to use any of the journal APIs first.
> 
> What do you think?

I think it will depend on how much from the journal other than subject will be need to be passed through, whilst separating out a few items makes intent clear, passing through too many items tends to make it more difficult to read/follow.

oliverfoster commented:
[heart]         Oliver Foster reacted to your message:
________________________________
From: joe-allen-89 ***@***.***>
Sent: Thursday, March 14, 2024 11:26:51 AM
To: cgkineo/adapt-migrations ***@***.***>
Cc: Oliver Foster ***@***.***>; Mention ***@***.***>
Subject: Re: [cgkineo/adapt-migrations] Fix: Pass data, fromPlugins, toPlugins, originalPlugins through journal rather than separately (part of #4) (PR #11)


@oliverfoster<https://github.com/oliverfoster> I've gone through and changed data to content where needed. I've also updated the function names accordingly to show that they're dealing with content rather than data.

—
Reply to this email directly, view it on GitHub<https://github.com/cgkineo/adapt-migrations/pull/11#issuecomment-1997230938>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB426B3ARAL53URB42DJFBDYYGCPXAVCNFSM6AAAAABER34EA6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJXGIZTAOJTHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
